### PR TITLE
Robot header

### LINKED ISSUE #1152: Add option to read in template information from a separate external file
Currently ROBOT template files as described  [here](http://robot.obolibrary.org/template) have to include the template strings for each column as the second line in the input template file. 
An option to add the ability to read the template string definitions from a separate file would allow for a more flexible pipeline where the template strings can be separately created and input via a file. 
A usage example would be
`robot template --template animals_ext_template.tsv \`    
      `--ext-template animals_template.tsv \`    
      `--output results/animals_ext_template.owl`

### PR DESCRIPTION
**Pull request by pk-mitre**
Resolves #1152

- [x] `docs/` have been added/updated
- [x] tests have been added/updated
- [x] `mvn verify` says all tests pass
- [x] `mvn site` says all JavaDocs correct
- [x] `CHANGELOG.md` has been updated

Adds option to input template strings via an external file as described in Issue #1152

### CODE REVIEW

**Comment by jamesaoverton** [comment]
My first thought when I see the details of this implementation is that the row numbers in our error messages will be off by one, which will be confusing.

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+-- Added option to input template strings from external file [##1152]
```
Comment (jamesaoverton) [inline CHANGELOG.md:11]: `##1152` should be just `#1152`

This link needs an expansion at the bottom of the file.

File: docs/examples/animals_template_combined.tsv [hunk 2]
```diff
@@ -0,0 +1,5 @@
+CURIE	Label	Parent	Comment
+ID	LABEL	SC %	A rdfs:comment
+obo:0000001	animal		Any animal in the world.
+0000002	canine	animal	A member of the genus Canis.
```
Comment (jamesaoverton) [inline docs/examples/animals_template_combined.tsv:4]: Is this combined file supposed to include this error? The file doesn't seem to be used anywhere, so I would not include it.

File: robot-core/src/main/java/org/obolibrary/robot/TemplateOperation.java [hunk 3]
```diff
@@ -158,7 +159,7 @@ public static OWLOntology template(
    * Given an OWLOntology, an IOHelper, a map of table names to table contents, and a map of
    * template options, use the tables as templates to generate a merged output ontology.
    *
-   * @param inputOntology OWLOntology to use to get existing entities
+   * @param inputOntology OWLOntology to use to get existing entitfies
```
Comment (jamesaoverton) [inline robot-core/src/main/java/org/obolibrary/robot/TemplateOperation.java:162]: Should be "entities"

File: robot-core/src/main/java/org/obolibrary/robot/TemplateOperation.java [hunk 4]
```diff
@@ -487,6 +492,8 @@ public static OWLOntology template(
       int idColumn = idColumns.get(tableName);
       int labelColumn = labelColumns.get(tableName);
       List<List<String>> rows = tables.get(tableName);
+      // TODO The starting row has to be controlled by the approach used, ie, internal or external
```
Comment (jamesaoverton) [inline robot-core/src/main/java/org/obolibrary/robot/TemplateOperation.java:495]: Is this TODO now done? If so the comment should be removed.

File: docs/examples/animals_template_combined.tsv (hunk 2 above)
Comment (jamesaoverton) [reply docs/examples/animals_template_combined.tsv:4]: To answer my own question: Yes, I think this file is included to compare row numbers for errors. So I suggest naming this file `animals_template_error.tsv` to match `animals_ext_template_error.tsv`.

**Comment by jamesaoverton** [comment]
@pk-mitre I apologize for taking so long to review this PR. This is a good feature, and we really do appreciate your contribution! The problem is simply that nobody on the core ROBOT team has dedicated hours for this work.

I made a few specific suggestions. If you have time to reply or make those changes, it would be appreciated. If you don't have time, let me know and I'll try to find time to finish the PR myself.

The only major question I have at this point is: What to do if the template and the ext-template do not match? Eventually someone will add new columns to one and forget to add them to the other, then file a bug report here.

I think that that best approach might be to check that the first row of the template and the ext-template are identical. What do you think?

Thanks again!

**Comment by pk-mitre** [comment]
James:
I have addressed the code comments and updated the pull request code.
Thanks

**Comment by jamesaoverton** [comment]
@pk-mitre Thank you for your continued work on this PR. I'm sorry that our schedules never seem to be in sync.

Do you still plan to follow up on the case of a mismatch between the template and the ext-template? Please let me know either way. Thanks again!

**Comment by pk-mitre** [comment]
@jamesaoverton  Yes, sorry I was working on a few other things. I will be following up on this mismatch issue and check things in.
Thanks !

**Comment by pk-mitre** [comment]
@jamesaoverton Am working through any potential additional error handling. With reference to your suggestion about checking the first row of the template file against the external template,I assume you mean the column count ? Since the first row in the template file is the header information, so it wouldn't be identical with the externalized template strings even in a correct case. For example:

Template file:

CURIE	Label	Parent	Comment
obo:0000001	animal		Any animal in the world.
obo:0000002	canine	animal	A member of the genus Canis.
obo:0000003	feline	animal	A member of the genus Felis.

external template file:

ID	LABEL	SC %	A rdfs:comment

File: docs/template.md [hunk 5]
```diff
@@ -223,6 +223,21 @@ robot template --merge-before \
   --output results/test_template.owl
 ```
 
+ROBOT template data read from separate external file
+
+    robot template --template animals_ext_template.tsv \
+        --ext-template animals_template.tsv \
```
Comment (jamesaoverton) [inline docs/template.md:229]: I think it's confusing that `--template` uses `animals_ext_template.tsv` while `--ext-template` uses `animals_template.tsv`. The names of these files should be swapped.

File: robot-command/src/main/java/org/obolibrary/robot/TemplateCommand.java [hunk 6]
```diff
@@ -49,6 +53,7 @@ public TemplateCommand() {
         "A", "include-annotations", true, "if true, include ontology annotations from merge input");
     o.addOption("f", "force", true, "if true, do not exit on error");
     o.addOption("e", "errors", true, "write errors to this path (TSV or CSV)");
+    o.addOption("E", "ext-template", true, "external robot template data file");
```
Comment (jamesaoverton) [inline robot-command/src/main/java/org/obolibrary/robot/TemplateCommand.java:56]: I'd prefer to call this `external-template` without abbreviating the name of the option.

**Comment by pk-mitre** [comment]
@jamesaoverton I have made the changes for the external template header error checking and also the file name changes to make it clearer as indicated in your comment on the naming.

**Comment by jamesaoverton** [comment]
I'm going to merge this and make a new PR with a few changes.

Sorry that this one took so long. Thanks for hanging in there and getting it done!

**Comment by dlutz2** [comment]
@jamesaoverton Thanks for the merge, appreciate the work and the time.

**Comment by pk-mitre** [comment]
@jamesaoverton Thanks ! Glad it all merged in smoothly.

**Comment by pk-mitre** [comment]
@jamesaoverton This one line has to be updated to take into account the changed flag name:    
` if (options.get("external-template") != null) {
        template.setRowNum(1);
      }`
File: TemplateOperation.java

**Comment by jamesaoverton** [comment]
You're right. I fixed that in dc04a3c97b89c1306245479740cab50d615342f0. I really wish we had a test to catch that, but I ran into a bunch of problems verifying this, and used up more time than I could afford.

**Comment by pk-mitre** [comment]
@jamesaoverton Thanks for all your help on this ticket !

**Pull request merged**